### PR TITLE
Automated cherry pick of #123809: fix stateful set pod recreation and event spam

### PR DESCRIPTION
**Pull request by atiratree**
Cherry pick of #123809 on release-1.28.

#123809: fix stateful set pod recreation and event spam

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

```release-note

```

### CODE REVIEW

**Comment by atiratree** [comment]
added `getErr` method to the `requestTracker` for the unit tests to pass

**Comment by atiratree** [comment]
/hold
https://github.com/kubernetes/kubernetes/pull/124563#issuecomment-2079560446

**Review by soltysh — APPROVED** [review]
/approve

**Comment by atiratree** [comment]
first commit from https://github.com/kubernetes/kubernetes/pull/124624 also included to fix the flaking TestStatefulSetControl tests

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Review by Verolop — COMMENTED** [review]
/lgtm 
(release managers)

**Pull request merged**
